servicetalk-http-api/src/main/java/io/servicetalk/http/api/ContentEncodingHttpServiceFilter.java

```diff
@@ -0,0 +1,224 @@
+                                decoder.streamingDecoder().deserialize(pub, ctx.executionContext().bufferAllocator()));
+                    } else {
+                        requestDecompressed = request;
+                    }
+
+                    return super.handle(ctx, requestDecompressed, responseFactory).map(response -> {
+                        final CharSequence reqAcceptEncoding;
+                        if (isPassThrough(request.method(), response) ||
+                                (reqAcceptEncoding = request.headers().get(ACCEPT_ENCODING)) == null) {
+                            return response;
+                        }
+
+                        BufferEncoder encoder = negotiateAcceptedEncodingRaw(reqAcceptEncoding, compressors,
+                                BufferEncoder::encodingName);
+                        if (encoder == null || identityEncoder().equals(encoder)) {
+                            return response;
+                        }
+
+                        HeaderUtils.addContentEncoding(response.headers(), encoder.encodingName());
+                        return response.transformPayloadBody(bufPub ->
+                                encoder.streamingEncoder().serialize(bufPub, ctx.executionContext().bufferAllocator()));
+                    }).subscribeShareContext();
+                });
+            }
+        };
+    }
+
+    @Override
+    public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+        // No influence - no blocking
+        return strategy;
+    }
+
+    private static boolean isPassThrough(final HttpRequestMethod method, final StreamingHttpResponse response) {
+        // see. https://tools.ietf.org/html/rfc7230#section-3.3.3
+        // The length of a message body is determined by one of the following
+        //         (in order of precedence):
+        //
+        // 1.  Any response to a HEAD request and any response with a 1xx
+        //         (Informational), 204 (No Content), or 304 (Not Modified) status
+        // code is always terminated by the first empty line after the
+        // header fields, regardless of the header fields present in the
+        // message, and thus cannot contain a message body.
+        //
+        // 2.  Any 2xx (Successful) response to a CONNECT request implies that
+        // the connection will become a tunnel immediately after the empty
+        // line that concludes the header fields.  A client MUST ignore any
+        // Content-Length or Transfer-Encoding header fields received in
+        // such a message.
+        // ...
+        final int code = response.status().code();
+        return INFORMATIONAL_1XX.contains(code) || code == NO_CONTENT.code() || code == NOT_MODIFIED.code() ||
+                (method == HEAD || (method == CONNECT && SUCCESSFUL_2XX.contains(code)));
+    }
+
+    @Nullable
+    static <T> T matchAndRemoveEncoding(final List<T> supportedEncoders,
```

Comment:
I tried sharing code between these two methods but they are doing different jobs and I couldn't find a good way. I considered at least moving them into the same `HeaderUtils` file, to keep similar code close but the `matchAndRemoveEncoding` can be kept package private if it stays here.

`matchAndRemoveEncoding` -> needs to remove the encoding header that is matched at preserve ordering (bcz the body has been decoded and the header no longer applies). first header must supported values or else we can't decode and its an error.
`negotiateAcceptedEncodingRaw` -> finds a matching encoding, so it can be applied (no removal of headers). any match is acceptable.